File: mobile/components/TreatCard.js
// TreatCard is the Overview face of the earn-your-treats habit. It shows one
// active treat, its progress dots, and a one-tap check-in, so the habit lives
// where the user already looks every day. Tapping the body opens the full
// manage screen. When no treat is set it becomes a slim invite instead. It
// reuses the WeekChain look, and turns celebrate colored when a treat is
// earned. When payday is near it nudges to earn the treat before the money
// lands. No dashes, no health data, never blocks a purchase.

import { useMemo } from 'react';
import { Pressable, StyleSheet, Text, View } from 'react-native';
import { useRouter } from 'expo-router';
import { Ionicons } from '@expo/vector-icons';
import { spacing, radius, fontSize, fontWeight } from '../theme';
import { useTheme } from '../context/Theme';
import { useAppData } from '../context/AppData';
import { treatStatus, toggleCheckIn } from '../lib/treats';
import { daysUntilPayday } from '../lib/format';

export default function TreatCard() {
  const { colors } = useTheme();
  const styles = useMemo(() => makeStyles(colors), [colors]);
  const router = useRouter();
  const { data, updateSettings } = useAppData();

  const treats = Array.isArray(data.settings && data.settings.treats) ? data.settings.treats : [];

  // Show the treat that most needs attention: the first one not yet earned,
  // else the first earned one so the win still shows.
  const statuses = treats.map((t) => treatStatus(t));
  const pick = statuses.find((s) => !s.earned) || statuses[0] || null;

  const onToggle = (id) => {
    updateSettings((s) => ({
      treats: (Array.isArray(s.treats) ? s.treats : []).map((t) => (t.id === id ? toggleCheckIn(t) : t)),
    }));
  };

  // Slim invite when no treat exists yet, so the feature is discoverable
  // without crowding Overview.
  if (!pick) {
    return (
      <Pressable
        onPress={() => router.push('/treats')}
        style={({ pressed }) => [styles.invite, pressed && styles.pressed]}
      >
        <Text style={styles.inviteEmoji}>☕</Text>
        <View style={{ flex: 1 }}>
          <Text style={styles.inviteTitle}>Earn your treats</Text>
          <Text style={styles.inviteSub}>Pair a small reward with a healthy habit. Guilt free.</Text>
        </View>
        <Ionicons name="chevron-forward" size={16} color={colors.faint} />
      </Pressable>
    );
  }

  // Only nudge on payday when the user actually set a schedule. Without one
  // the default would fire the nudge on random days and read as noise.
  const schedule = data.settings && data.settings.paydaySchedule;
  const paydaySoon = !!schedule && daysUntilPayday(new Date(), schedule) <= 1;

  const subtitle = pick.earned
    ? `Earned. Enjoy your ${String(pick.treat || 'treat').toLowerCase()}, you paid for it in ${String(pick.action || 'effort').toLowerCase()}.`
    : paydaySoon
    ? `Sweldo malapit na. Lock in your ${String(pick.action || 'habit').toLowerCase()} habit first, ${pick.remaining} to go.`
    : pick.recent === 0
    ? `Do your ${String(pick.action || 'habit').toLowerCase()}, then check in. ${pick.target} earns it.`
    : `${pick.recent} of ${pick.target} check ins. ${pick.remaining} more and it is yours.`;

  return (
    <View style={[styles.card, pick.earned && styles.cardEarned]}>
      <Pressable onPress={() => router.push('/treats')} style={styles.head} hitSlop={4}>
        <Text style={styles.kicker}>EARN YOUR TREAT</Text>
        <Ionicons name="chevron-forward" size={16} color={colors.faint} />
      </Pressable>

      <Pressable onPress={() => router.push('/treats')} style={styles.body}>
        <Text style={styles.emoji} importantForAccessibility="no">{pick.emoji}</Text>
        <View style={{ flex: 1 }}>
          <Text style={styles.treat}>{pick.treat}</Text>
          <View style={styles.dotsRow}>
            {Array.from({ length: pick.target }).map((_, i) => (
              <View
                key={i}
                style={[
                  styles.dot,
                  i < pick.recent && styles.dotOn,
                  pick.earned && i < pick.recent && styles.dotEarned,
                ]}
              />
            ))}
          </View>
        </View>
        {pick.earned ? <Text style={styles.earnedTag}>EARNED</Text> : null}
      </Pressable>

      <Text style={[styles.sub, pick.earned && styles.subEarned]}>{subtitle}</Text>

      <Pressable
        onPress={() => onToggle(pick.id)}
        style={({ pressed }) => [styles.checkBtn, pick.doneToday && styles.checkBtnDone, pressed && styles.pressed]}
      >
        <Ionicons
          name={pick.doneToday ? 'checkmark-circle' : 'ellipse-outline'}
          size={18}
          color={pick.doneToday ? colors.background : colors.primary}
        />
        <Text style={[styles.checkBtnText, pick.doneToday && styles.checkBtnTextDone]}>
          {pick.doneToday ? 'Done for today, tap to undo' : 'I did it today'}
        </Text>
      </Pressable>
    </View>
  );
}

function makeStyles(colors) {
  return StyleSheet.create({
    card: {
      backgroundColor: colors.card,
      borderColor: colors.border,
      borderWidth: 1,
      borderRadius: radius.lg,
      padding: spacing.xl,
      marginBottom: spacing.lg,
    },
    cardEarned: { borderColor: colors.celebrate || colors.primary },
    head: { flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between' },
    kicker: { color: colors.softGreen, fontSize: fontSize.caption, fontWeight: fontWeight.medium, letterSpacing: 1.2 },

    body: { flexDirection: 'row', alignItems: 'center', gap: spacing.md, marginTop: spacing.md },
    emoji: { fontSize: 26 },
    treat: { color: colors.text, fontSize: fontSize.body, fontWeight: fontWeight.bold },
    dotsRow: { flexDirection: 'row', flexWrap: 'wrap', gap: spacing.sm, marginTop: spacing.sm },
    dot: { width: 14, height: 14, borderRadius: 7, borderWidth: 1.5, borderColor: colors.border },
    dotOn: { backgroundColor: colors.primary, borderColor: colors.primary },
    dotEarned: { backgroundColor: colors.celebrate || colors.primary, borderColor: colors.celebrate || colors.primary },
    earnedTag: {
      color: colors.celebrate || colors.primary,
      fontSize: 10,
      fontWeight: fontWeight.bold,
      letterSpacing: 1,
      borderColor: colors.celebrate || colors.primary,
      borderWidth: 1,
      borderRadius: radius.sm,
      paddingHorizontal: 6,
      paddingVertical: 1,
    },

    sub: { color: colors.textSecondary, fontSize: fontSize.small, lineHeight: 19, marginTop: spacing.md },
    subEarned: { color: colors.celebrate || colors.primary, fontWeight: fontWeight.medium },

    checkBtn: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      gap: spacing.sm,
      borderColor: colors.primary,
      borderWidth: 1,
      borderRadius: radius.pill,
      paddingVertical: spacing.md,
      marginTop: spacing.md,
    },
    checkBtnDone: { backgroundColor: colors.primary },
    checkBtnText: { color: colors.primary, fontSize: fontSize.small, fontWeight: fontWeight.bold },
    checkBtnTextDone: { color: colors.background },

    invite: {
      flexDirection: 'row',
      alignItems: 'center',
      gap: spacing.md,
      backgroundColor: colors.card,
      borderColor: colors.border,
      borderWidth: 1,
      borderRadius: radius.lg,
      padding: spacing.lg,
      marginBottom: spacing.lg,
    },
    inviteEmoji: { fontSize: 24 },
    inviteTitle: { color: colors.text, fontSize: fontSize.body, fontWeight: fontWeight.bold },
    inviteSub: { color: colors.muted, fontSize: fontSize.caption, marginTop: 2 },

    pressed: { opacity: 0.7 },
  });
}

File: flutter/lib/widgets/treat_card.dart
// The treat surfaced where the habit lives: Home. The full Treats screen
// existed but the daily loop (do the healthy thing, check in, earn the
// treat) was out of sight behind Menu, which is where habits go to die.
// Ported from the RN TreatCard: picks the first not-yet-earned treat (or
// the first earned one so the win still shows), progress dots, and the
// one-tap check-in writing through the same store method the Treats
// screen uses.
//
// The treat emoji is USER DATA (they picked it), so it renders as-is; the
// icon rule only governs Salapify-authored icons.

import 'package:flutter/material.dart';

import '../data/store.dart';
import '../money/schedule.dart' show daysUntilPayday;
import '../money/treats.dart' as treats;
import '../screens/treats.dart' show TreatsScreen;
import '../theme.dart';
import '../typography.dart';
import '../widgets/salapify_icon.dart';

class TreatCard extends StatelessWidget {
  final SalapifyStore store;

  /// Injectable clock, the usual seam.
  final DateTime Function() clock;

  // ignore: prefer_const_constructors_in_immutables
  TreatCard({super.key, required this.store, this.clock = DateTime.now});

  void _open(BuildContext context) => Navigator.of(
    context,
  ).push(MaterialPageRoute(builder: (_) => TreatsScreen(store: store)));

  /// The quiet one-line form of this card: title, one sub line, a chevron,
  /// the same tap through to Treats. Used when there is nothing yet and when
  /// the pick is not close enough to earn front-page excitement.
  Widget _slimRow(
    BuildContext context, {
    required String title,
    required String sub,
  }) {
    return Card(
      child: Semantics(
        button: true,
        label: title,
        child: InkWell(
          borderRadius: BorderRadius.circular(12),
          onTap: () => _open(context),
          child: Padding(
            padding: const EdgeInsets.all(16),
            child: Row(
              children: [
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(title, style: AppText.bodyStrong),
                      const SizedBox(height: 2),
                      Text(sub, style: AppText.caption.copyWith(height: 1.35)),
                    ],
                  ),
                ),
                ExcludeSemantics(
                  child: Icon(
                    salapifyIcon('forward'),
                    size: 18,
                    color: Barako.muted,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final ref = clock();
    final statuses = [
      for (final t in store.treatRules) treats.treatStatus(t, ref),
    ];
    Map<String, dynamic>? pick;
    for (final s in statuses) {
      if (s['earned'] != true) {
        pick = s;
        break;
      }
    }
    pick ??= statuses.isNotEmpty ? statuses.first : null;

    if (pick == null) {
      // No treats yet: a slim invite, not a wall.
      return _slimRow(
        context,
        title: 'Earn your treats',
        sub: 'Pair a small reward with a healthy habit. Guilt free.',
      );
    }

    final earned = pick['earned'] == true;
    final doneToday = pick['doneToday'] == true;
    final target = pick['target'] as int;
    final recent = pick['recent'] as int;
    final remaining = pick['remaining'] as int;
    final action = (pick['action'] ?? 'habit').toString().toLowerCase();
    final treatName = (pick['treat'] ?? 'treat').toString();

    // The full card with its dots earns front-page space only when something
    // is HAPPENING: the treat is earned (the win must show) or one check-in
    // away (the nudge can close it today). Mid-journey, one quiet line keeps
    // the habit visible without spending ~140dp of Home on it every day.
    if (!earned && remaining > 1) {
      return _slimRow(
        context,
        title: 'Earn your treats',
        sub:
            '$recent of $target toward ${treatName.toLowerCase()}. '
            '${remaining == target ? 'Start today.' : 'Keep going.'}',
      );
    }

    // Payday nudge only when the user explicitly set a schedule; without one
    // the line would fire on guesses and read as noise, the RN rule.
    final schedule = (store.data['settings'] is Map)
        ? (store.data['settings'] as Map)['paydaySchedule']
        : null;
    final paydaySoon = schedule != null && daysUntilPayday(ref, schedule) <= 1;

    final String sub;
    if (earned) {
      sub =
          'Earned. Enjoy your ${treatName.toLowerCase()}, you paid for it '
          'in $action.';
    } else if (paydaySoon) {
      sub =
          'Payday is close. Lock in your $action habit first, $remaining '
          'to go.';
    } else if (recent == 0) {
      sub = 'Do your $action, then check in. $target earns it.';
    } else {
      sub = '$recent of $target check ins. $remaining more and it is yours.';
    }

    return Card(
      shape: earned
          ? RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(Radii.card),
              side: BorderSide(color: Barako.celebrate),
            )
          : null,
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Semantics(
              button: true,
              label: 'Earn your treat, open Treats',
              child: InkWell(
                borderRadius: BorderRadius.circular(8),
                onTap: () => _open(context),
                child: Row(
                  children: [
                    Text('EARN YOUR TREAT', style: Barako.cardKickerStyle),
                    const Spacer(),
                    if (earned)
                      Container(
                        padding: const EdgeInsets.symmetric(
                          horizontal: 6,
                          vertical: 1,
                        ),
                        decoration: BoxDecoration(
                          border: Border.all(color: Barako.celebrate),
                          borderRadius: BorderRadius.circular(Radii.control),
                        ),
                        child: Text(
                          'EARNED',
                          style: TextStyle(
                            color: Barako.celebrate,
                            fontSize: 10,
                            fontWeight: FontWeight.w700,
                            letterSpacing: 1,
                          ),
                        ),
                      )
                    else
                      ExcludeSemantics(
                        child: Icon(
                          salapifyIcon('forward'),
                          size: 18,
                          color: Barako.muted,
                        ),
                      ),
                  ],
                ),
              ),
            ),
            const SizedBox(height: Gap.sm),
            Row(
              children: [
                if ((pick['emoji'] ?? '').toString().isEmpty)
                  SalapifyGlyph('treat', size: 22, boxed: false)
                else
                  Text(
                    (pick['emoji'] ?? '').toString(),
                    style: const TextStyle(fontSize: 24),
                  ),
                const SizedBox(width: Gap.sm),
                Expanded(child: Text(treatName, style: AppText.bodyStrong)),
              ],
            ),
            const SizedBox(height: Gap.sm),
            Wrap(
              spacing: 8,
              runSpacing: 8,
              children: [
                for (var i = 0; i < target; i++)
                  Container(
                    width: 14,
                    height: 14,
                    decoration: BoxDecoration(
                      shape: BoxShape.circle,
                      color: i < recent
                          ? (earned ? Barako.celebrate : Barako.primary)
                          : null,
                      border: Border.all(
                        color: i < recent
                            ? (earned ? Barako.celebrate : Barako.primary)
                            : Barako.border,
                        width: 1.5,
                      ),
                    ),
                  ),
              ],
            ),
            const SizedBox(height: Gap.sm),
            Text(
              sub,
              style: AppText.small.copyWith(
                color: earned ? Barako.celebrate : Barako.textSecondary,
                height: 1.4,
                fontWeight: earned ? TypeWeight.medium : TypeWeight.regular,
              ),
            ),
            // The check-in button: for an unearned treat, or to undo today's
            // check-in on an earned one. Never without a writable store.
            if (store.canWrite && (!earned || doneToday)) ...[
              const SizedBox(height: Gap.md),
              SizedBox(
                width: double.infinity,
                child: doneToday
                    ? FilledButton.icon(
                        onPressed: () => store.toggleTreatCheckIn(
                          (pick!['id'] ?? '').toString(),
                        ),
                        icon: Icon(salapifyIcon('selected'), size: 18),
                        label: const Text('Done for today, tap to undo'),
                        style: FilledButton.styleFrom(
                          backgroundColor: Barako.primary,
                          foregroundColor: Barako.onPrimary,
                          minimumSize: const Size(0, 48),
                        ),
                      )
                    : OutlinedButton.icon(
                        onPressed: () => store.toggleTreatCheckIn(
                          (pick!['id'] ?? '').toString(),
                        ),
                        icon: Icon(salapifyIcon('unselected'), size: 18),
                        label: const Text('I did it today'),
                        style: OutlinedButton.styleFrom(
                          foregroundColor: Barako.primaryText,
                          side: BorderSide(color: Barako.primary),
                          minimumSize: const Size(0, 48),
                        ),
                      ),
              ),
            ],
          ],
        ),
      ),
    );
  }
}
